holoviews/plotting/bokeh/stats.py

```diff
@@ -342,18 +342,52 @@ class ViolinPlot(BoxWhiskerPlot):
     selection_display = BokehOverlaySelectionDisplay(color_prop='violin_fill_color')
 
     def _kde_data(self, el, key, **kwargs):
-        vdim = el.vdims[0]
-        values = el.dimension_values(vdim)
+        vdims = el.vdims
+        vdim = vdims[0]
```

Comment:
-1 on using vdims for this, HoloViews is almost entirely built around the idea that the data is organized in a tidy format. It'd have to be an additional kdim which can optionally be mapped using a plot option.